metadata/src/main/java/org/apache/kafka/timeline/SnapshottableHashTable.java

```diff
@@ -0,0 +1,449 @@
+                 * slot 0 in the snapshot tier.
+                 * Elements that would be in slot 2 or 3 in the top tier can only be in
+                 * slot 1 in the snapshot tier.
+                 *
+                 * Therefore, we can do something like this:
+                 * 1. check slot 0 in the top tier and slot 0 in the snapshot tier.
+                 * 2. check slot 1 in the top tier and slot 0 in the snapshot tier.
+                 * 3. check slot 2 in the top tier and slot 1 in the snapshot tier.
+                 * 4. check slot 3 in the top tier and slot 1 in the snapshot tier.
+                 *
+                 * If elements move from the top tier to the snapshot tier, then
+                 * we'll still find them and report them exactly once.
+                 *
+                 * Note that while I used 4 and 2 as example sizes here, the same pattern
+                 * holds for different powers of two.  The "snapshot slot" of an element
+                 * will be the top few bits of the top tier slot of that element.
+                 */
+                Iterator<Snapshot> iterator = snapshotRegistry.iterator(snapshot);
+                while (iterator.hasNext()) {
+                    Snapshot curSnapshot = iterator.next();
+                    HashTier<T> tier = curSnapshot.getDelta(SnapshottableHashTable.this);
+                    if (tier != null && tier.deltaTable != null) {
+                        BaseHashTable<T> deltaTable = tier.deltaTable;
+                        int shift = Integer.numberOfLeadingZeros(deltaTable.baseElements().length) -
+                            Integer.numberOfLeadingZeros(topTier.length);
+                        int tierSlot = slot >>> shift;
+                        BaseHashTable.unpackSlot(temp, deltaTable.baseElements(), tierSlot);
+                        for (T object : temp) {
+                            if (BaseHashTable.findSlot(object, topTier.length) == slot) {
+                                ready.add(object);
+                            }
+                        }
+                        temp.clear();
+                    }
+                }
+                slot++;
+            }
+            return true;
+        }
+
+        @Override
+        public T next() {
+            if (!hasNext()) {
+                throw new NoSuchElementException();
+            }
+            return ready.remove(ready.size() - 1);
+        }
+    }
+
+    private final SnapshotRegistry snapshotRegistry;
+
+    SnapshottableHashTable(SnapshotRegistry snapshotRegistry, int expectedSize) {
+        super(expectedSize);
+        this.snapshotRegistry = snapshotRegistry;
+    }
+
+    int snapshottableSize(long epoch) {
+        if (epoch == LATEST_EPOCH) {
+            return baseSize();
+        } else {
+            Iterator<Snapshot> iterator = snapshotRegistry.iterator(epoch);
+            while (iterator.hasNext()) {
```

Comment:
It seems that we don't need to do the while loop.

Reply:
same